### ISSUE TITLE
chainer.grad computes gradients for variables that are requires_grad=False

### ISSUE THREAD

**Issue by hvy** [chainer/chainer#5974]
`chainer.grad` computes and returns the gradients for variables that are flagged with `requires_grad=False`. 

This could be reconsidered to return `None` since `None` is returned for variables that are not included in the graph for instance.

```python
import numpy as np
import chainer as ch

a = ch.Variable(np.arange(3, dtype='f'), requires_grad=True)
b = ch.Variable(np.arange(3, dtype='f'), requires_grad=False)
c = ch.Variable(np.arange(3, dtype='f'))  # Will not be included in graph.

d = a * b
y = ch.functions.sum(d)

ga, gb, gc = ch.grad([y], [a, b, c])

print(ga)  # variable([0. 1. 2.]), Expected
print(gb)  # variable([0. 1. 2.]), Unexpected and should be None
print(gc)  # None, Expected
```

**Comment by toslunar** [comment]
IMO, `grad` should raise an error for that case.

**Comment by toslunar** [comment]
Similarly to #5709, I'd say `chainer.grad` is buggy because the semantics depends on (the number of) `FunctionNode`s to implement a function.  In this example, `gb` will be `None` with
```
d = a * ch.functions.identity(b)
```

**Comment by toslunar** [comment]
The issue arose in chainer/chainerrl#510.
